File: DiaryNutrition/app/src/androidTest/java/ediger/diarynutrition/db/MealDaoTest.java
package ediger.diarynutrition.db;

import org.junit.After;
import org.junit.Before;
import org.junit.Rule;
import org.junit.Test;
import org.junit.runner.RunWith;

import java.util.List;

import androidx.arch.core.executor.testing.InstantTaskExecutorRule;
import androidx.room.Room;
import androidx.test.InstrumentationRegistry;
import androidx.test.runner.AndroidJUnit4;
import ediger.diarynutrition.LiveDataTestUtil;
import ediger.diarynutrition.data.DiaryDatabase;
import ediger.diarynutrition.data.Meal;
import ediger.diarynutrition.data.dao.MealDao;

import static org.hamcrest.MatcherAssert.assertThat;
import static org.hamcrest.CoreMatchers.notNullValue;
import static org.hamcrest.CoreMatchers.is;
import static org.junit.Assert.assertNull;


@RunWith(AndroidJUnit4.class)
public class MealDaoTest {

    static final Meal MEAL = new Meal("Breakfast");
    static final Meal MEAL2 = new Meal("Lunch");

    @Rule
    public InstantTaskExecutorRule instantTaskExecutorRule = new InstantTaskExecutorRule();

    private DiaryDatabase mDatabase;
    private MealDao mMealDao;

    @Before
    public void initDb() {
        mDatabase = Room.inMemoryDatabaseBuilder(InstrumentationRegistry.getContext(),
                DiaryDatabase.class)
                .allowMainThreadQueries()
                .build();
        mMealDao = mDatabase.mealDao();
    }

    @After
    public void closeDb() {
        mDatabase.close();
    }

    @Test
    public void insertAndGetById() {
        mMealDao.insertMeal(MEAL);
        Meal meal = mMealDao.getMeal(1);
        assertMeal(meal, MEAL.getName());
    }

    @Test
    public void insertAndUpdate() {
        mMealDao.insertMeal(MEAL);
        Meal meal = mMealDao.getMeal(1);
        meal.setName("Name");

        mMealDao.updateMeal(meal);
        meal = mMealDao.getMeal(1);
        assertMeal(meal, "Name");
    }

    @Test
    public void insertAndDeleteById() {
        mMealDao.insertMeal(MEAL);
        mMealDao.deleteMealById(1);
        Meal meal = mMealDao.getMeal(1);
        assertNull(meal);
    }

    @Test
    public void insertAndGetAll() throws Exception {
        mMealDao.insertMeal(MEAL);
        mMealDao.insertMeal(MEAL2);

        List<Meal> meals = LiveDataTestUtil.getValue(mMealDao.getMeals());
        assertThat(meals.size(), is(2));

        assertMeal(meals.get(1), MEAL2.getName());
    }

    private void assertMeal(Meal meal, String name) {
        assertThat(meal, notNullValue());
        assertThat(meal.getName(), is(name));
    }

}

File: DiaryNutrition/app/src/androidTest/java/ediger/diarynutrition/db/MealDaoTest.kt
package ediger.diarynutrition.db

import androidx.arch.core.executor.testing.InstantTaskExecutorRule
import androidx.room.Room
import androidx.test.core.app.ApplicationProvider
import androidx.test.ext.junit.runners.AndroidJUnit4
import ediger.diarynutrition.data.source.DiaryDatabase
import ediger.diarynutrition.data.source.entities.Meal
import kotlinx.coroutines.ExperimentalCoroutinesApi
import kotlinx.coroutines.test.runBlockingTest
import org.hamcrest.MatcherAssert.assertThat
import org.hamcrest.core.Is.`is`
import org.hamcrest.core.IsNull.notNullValue
import org.junit.After
import org.junit.Before
import org.junit.Rule
import org.junit.Test
import org.junit.runner.RunWith

fun Meal.assertMeal(original: Meal) {
    assertThat(this, notNullValue())
    assertThat(this.name, `is`(original.name))
}

@ExperimentalCoroutinesApi
@RunWith(AndroidJUnit4::class)
class MealDaoTest {

    @get:Rule
    var instantExecutorRule = InstantTaskExecutorRule()

    private lateinit var database: DiaryDatabase

    @Before
    fun initDb() {
        database = Room.inMemoryDatabaseBuilder(
                ApplicationProvider.getApplicationContext(),
                DiaryDatabase::class.java
        ).build()
    }

    @After
    fun closeDb() {
        database.close()
    }

    @Test
    fun insertAndGetById() = runBlockingTest {
        database.mealDao().insertMeal(MEAL)
        val meal = database.mealDao().getMeal(1)
        meal.assertMeal(MEAL)
    }

    @Test
    fun insertAndUpdate() = runBlockingTest {
        database.mealDao().insertMeal(MEAL)
        var meal = database.mealDao().getMeal(1).apply { name = "Different" }

        database.mealDao().updateMeal(meal)
        meal = database.mealDao().getMeal(1)

        meal.assertMeal(MEAL.apply { name = "Different" })
    }

    @Test
    fun insertAndDeleteById() = runBlockingTest {
        database.mealDao().insertMeal(MEAL)

        var mealList = database.mealDao().getMeals()
        assertThat(mealList.size, `is`(1))

        database.mealDao().deleteMealById(1)

        mealList = database.mealDao().getMeals()
        assertThat(mealList.size, `is`(0))
    }

    @Test
    fun insertAndGetAll() = runBlockingTest {
        database.mealDao().insertMeal(MEAL)
        database.mealDao().insertMeal(MEAL_2)

        val mealList = database.mealDao().getMeals()
        assertThat(mealList.size, `is`(2))

        mealList.find { it.name == MEAL.name }!!.assertMeal(MEAL)
        mealList.find { it.name == MEAL_2.name }!!.assertMeal(MEAL_2)
    }

    @Test
    fun populateMeals() = runBlockingTest {
        database.mealDao().populateMeals(listOf(MEAL, MEAL_2))

        val mealList = database.mealDao().getMeals()
        assertThat(mealList.size, `is`(2))

        mealList.find { it.name == MEAL.name }!!.assertMeal(MEAL)
        mealList.find { it.name == MEAL_2.name }!!.assertMeal(MEAL_2)
    }

    companion object {
        val MEAL = Meal("Breakfast", 0)
        val MEAL_2 = Meal("Lunch", 0)
    }
}